### PR TITLE
[TE/JAX] XLA FFI calls for Softmax and FusedAttnBackward

### PR DESCRIPTION
**Pull request by huanghua1994**
# Description

This PR introduced the following primitives implemented with the new custom calls:

* `ScaledSoftmaxFwdPrimitive`
* `ScaledSoftmaxBwdPrimitive`
* `ScaledMaskedSoftmaxFwdPrimitive`
* `ScaledMaskedSoftmaxBwdPrimitive`
* `ScaledUpperTriangMaskedSoftmaxFwdPrimitive`
* `ScaledUpperTriangMaskedSoftmaxBwdPrimitive`
* `FusedAttnBwdPrimitive`

Also added `DequantizeFFI()` in `transformer_engine/jax/csrc/extensions/quantization.cpp` although currently no Python function calls dequantize explicitly. 

All C++ functions in `transformer_engine/jax/csrc/extensions` have FFI after this PR.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by huanghua1994** [comment]
@zlsh80826 The forward test in `test_fused_attn.py` is turned on in this commit, and it takes about 15 minutes to run on a single H100. I think since we need to run L1 tests for any update related to fused attention, it is better to test both the forward and backward primitives in CI.

**Comment by huanghua1994** [comment]
CI L1 tests passed. Rebase to the main branch to verify #1314

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -735,7 +735,7 @@ class TestFusedAttn:
             pytest.param(False, id="INFERENCE"),
         ],
     )
-    def _test_forward(
+    def test_forward(
```
[the remark this answers is not in the document]
Comment (huanghua1994) [reply tests/jax/test_fused_attn.py:738]: I think `test_forward` is not really so time-consuming,  just about 15 minutes on a H100 (`test_backward` is about 10 minutes). If `test_backward` covers forward testing in `value_and_grad`, we don't need to have `test_forward` in CI.

[the remark this answers is not in the document]
Comment (zlsh80826) [reply tests/jax/test_fused_attn.py:738]: I prefer to just keep `test_backward` only here, 15 mins is still not a small time as the time will keep increased with the more attention variants get supported.

If you think we still need a subset for `test_forward` sanity check, another approach is that we can specialize 
```
@pytest.mark.parametrize(
    "b, s_q, s_kv, h_q, h_kv, d, dtype",
```
for both `test_forward` and `test_backward`, and use only one set for `test_forward`.

[the remark this answers is not in the document]
Comment (huanghua1994) [reply tests/jax/test_fused_attn.py]: Okay, I have reverted the changes to `test_fused_attn.py`.

**Review by zlsh80826 — APPROVED** [review]
LGTM for the fused attn

**Review by phu0ngng — APPROVED** [review]
LGTM

**Pull request merged**